crates/typst-library/src/visualize/image.rs

```diff
@@ -64,6 +68,60 @@ pub struct ImageElem {
     pub fit: ImageFit,
 }
 
+/// A raster or vector graphic loaded from string.
+///
+/// Supported formats are PNG, JPEG, GIF and SVG.
+///
+/// _Note:_ Work on SVG export is ongoing and there might be visual inaccuracies
+/// in the resulting PDF. Make sure to double-check embedded SVG images. If you
+/// have an issue, also feel free to report it on [GitHub][gh-svg].
+///
+/// ## Example { #example }
+/// ```example
+/// #figure(
+///   image.decode(read("diagram.svg").replace("yellow", "green"), "svg", width: 80%),
+///   caption: [
+///     Results of the molecular testing pipeline.
+///   ],
+/// )
+/// ```
+///
+/// [gh-svg]: https://github.com/typst/typst/issues?q=is%3Aopen+is%3Aissue+label%3Asvg
+///
+/// Display: Image
+/// Category: visualize
+#[func]
+pub fn image_decode(
+    /// Image content
+    content: EcoString,
+    /// Image format
+    format: EcoString,
```

Comment:
Upon further consideration, I think it would be good to make a named, optional parameter `Smart<ImageFormat>`. You can derive `Cast` on `RasterFormat` and `VectorFormat` in `typst/src/image.rs` and use `cast!` to make `ImageFormat` itself castable. 

For the auto case, the format should be detected automatically. Strings can always be interpreted as SVG (at least for now, if we add another plain text image format we might need to be smarter about this) and for binary images, you can use the `image` crate's detection functionality to infer the format from the raw data.

Reply:
i'll have a look soon. Any hint/documentation about the cast macro?

Reply:
There isn't really documentation, but there are lots of use cases across the codebase which should show you how it works.

Reply:
Ok. I also added the format parameter to ImageElem, to avoid converting Format -> ext -> Format. Thereby I also enabled auto_detection on regular image, if no valid extention is given.